### PR TITLE
Move local execution after remote execution

### PR DESCRIPTION
**Pull request by onderkalaci**
Before this commit, when both local and remote tasks
exist, the executor was starting the execution with
local execution. There is no strict requirements on
this.

Especially considering the adaptive connection management
improvements that we plan to roll soon, moving the local
execution after to the remote execution makes more sense.

The adaptive connection management for single node Citus
would look roughly as follows:

   - Try to connect back to the coordinator for running
     parallel queries.
        - If succeeds, go on and execute tasks in parallel
        - If fails, fallback to the local execution

So, we'll use local execution as a fallback mechanism. And,
moving it after to the remote execution allows us to implement
such further scenarios.

### CODE REVIEW

File: src/backend/distributed/executor/local_executor.c [hunk 1]
```diff
@@ -560,14 +561,26 @@ ExecuteLocalTaskPlan(PlannedStmt *taskPlan, char *queryString,
...
+	bool storeTuples = true;
 
 	RecordNonDistTableAccessesForTask(task);
 
+	/*
+	 * For modifications to reference tables (or any replicated table in
+	 * general), remote execution have already stored the tuples and
+	 * incremented es_processed, so skip now.
+	 */
+	if (task->partiallyLocalOrRemote && !ReadOnlyTask(task->taskType) &&
```
Comment (onderkalaci) [inline src/backend/distributed/executor/local_executor.c:573]: we already have tests with reference tables that triggers  this code=path. It'd be nice to have similar tests with replication > 1, one placement on  the coordinator the other is on the worker.

File: src/backend/distributed/executor/adaptive_executor.c [hunk 2]
```diff
@@ -768,24 +765,17 @@ AdaptiveExecutor(CitusScanState *scanState)
...
-		{
-			/*
-			 * For reference tables we already add rowsProcessed on the local execution,
-			 * this is required to ensure that mixed local/remote executions reports
-			 * the accurate number of rowsProcessed to the user.
-			 */
-			executorState->es_processed += execution->rowsProcessed;
-		}
+	CmdType commandType = job->jobQuery->commandType;
+	if (commandType != CMD_SELECT)
```
Comment (onderkalaci) [inline src/backend/distributed/executor/adaptive_executor.c:776]: This has been forever like this such that we don't set `executorState->es_processed ` for select queries. I couldn't find the reason yet.

Also, would that work fine with modifying ctes with SELECT query?

Comment (marcocitus) [reply src/backend/distributed/executor/adaptive_executor.c:776]: iirc, it gives an unexpected extra output in psql if you set it. (like `SELECT 3`)

File: src/backend/distributed/executor/local_executor.c (hunk 1 above)
Comment (marcocitus) [reply src/backend/distributed/executor/local_executor.c:573]: Why do we have the !ReadOnlyTask distinction?

Comment (onderkalaci) [reply src/backend/distributed/executor/local_executor.c:573]: Maybe results should always come from local execution for modifications as well

Comment (onderkalaci) [reply src/backend/distributed/executor/local_executor.c]: @marcocitus as we talked, now whenever the task is `partiallyLocalOrRemote`, we store the tuples from the local execution.

Well, it turns out it has always been the case. But, due to the bug in `ReceiveResults ()`, we ended up with weird hacks. Please consider the change in `ReceiveResults ()` as well.

Overall, the code seems much cleaner to me now. Let's see what you think.

File: src/backend/distributed/executor/adaptive_executor.c [hunk 3]
```diff
@@ -845,29 +834,25 @@ RunLocalExecution(CitusScanState *scanState, DistributedExecution *execution)
...
-	execution->tasksToExecute = execution->remoteTaskList;
+	if (list_length(execution->localTaskList) > 0)
+	{
+		/* we only need to execute the remote tasks */
+		execution->tasksToExecute = execution->remoteTaskList;
 
-	execution->totalTaskCount = list_length(execution->remoteTaskList);
-	execution->unfinishedTaskCount = list_length(execution->remoteTaskList);
+		execution->totalTaskCount = list_length(execution->remoteTaskList);
+		execution->unfinishedTaskCount = list_length(execution->remoteTaskList);
```
Comment (marcocitus) [inline src/backend/distributed/executor/adaptive_executor.c:854]: could we just assign the right values immediately in CreateDistributedExecution?

Comment (onderkalaci) [reply src/backend/distributed/executor/adaptive_executor.c]: Yes, makes sense.

To make is nicer, I did a small refactoring as well, which I wanted to do for a while now.

Basically, made the `tasksToExecute` field more explicit, and used `remoteTaskList` in the remote execution and `localTaskList` in the local execution. 

My vague memory tells me that you didn't think we need such a change, but please consider this.

Comment (onderkalaci) [reply src/backend/distributed/executor/adaptive_executor.c]: And, also realized that we should do the same change (e.g., move local execution after remote execution) for `ExecuteTaskListExtended`, that's why the refactoring seems required even more.

**Review by marcocitus — APPROVED** [review]

**Comment by metdos** [comment]
Do we need a description for this PR?

**Comment by onderkalaci** [comment]
> Do we need a description for this PR?

I think no because this is an internal change, so nothing visible to the user.

**Pull request merged**